src/oop/special_class_method/hard/q12/MainTest.java

```diff
@@ -0,0 +1,257 @@
+package oop.special_class_method.hard.q12;
+
+import global.ObjectTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.stream.Stream;
+
+import static global.tools.CustomAssertions._assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MainTest {
+    // Java
+
+    private final String className = "Laptop";
+    private final String attributeName1 = "yearMade";
+    private final String getAttributeMethodName1 = "getYearMade";
+    private final String setAttributeMethodName1 = "setYearMade";
+    private final String attributeName2 = "colour";
+    private final String getAttributeMethodName2 = "getColour";
+    private final String setAttributeMethodName2 = "setColour";
+    private ObjectTest testClass;
+
+    @BeforeEach
+    public void setUp() throws Throwable {
+        String classString = "oop.special_class_method.hard.q12." + className;
+        this.testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void laptopClassHasCorrectAttributes() {
+        assertTrue(testClass.hasField(attributeName1),
+                "Your " + className + " class is missing the " + attributeName1 + " attribute, or it is spelled incorrectly.");
+        assertTrue(testClass.hasField(attributeName1, int.class),
+                "Your " + attributeName1 + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(attributeName1, "private"),
+                "Your " + attributeName1 + " does not have the correct visibility modifier.");
+        assertTrue(testClass.hasField(attributeName2),
+                "Your " + className + " class is missing the " + attributeName2 + " attribute, or it is spelled incorrectly.");
+        assertTrue(testClass.hasField(attributeName2, String.class),
+                "Your " + attributeName2 + " does not have the correct type.");
+        assertTrue(testClass.hasModifier(attributeName2, "private"),
+                "Your " + attributeName2 + " does not have the correct visibility modifier.");
+    }
+
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of(2002, "Red"),
```

Comment:
Should add 0 and "" as test case as you use this for the methods work together test.